### PR TITLE
Defer chain lock to caller to allow atomic lock

### LINKED ISSUE #368: Ensure using deadlock avoidance mechanism on mutex acquiring
#### What happened:

Currently acquire more than one recursive mutex is ordered, that means acquiring same mutexes in other thread in different / reverse order cause a deadlock 

#### What you expected to happen:
No deadlock on reverse order mutex acquiring

#### How to reproduce it (as minimally and precisely as possible):

```
Time |      Thread A      |     Thread B
t1   | Attempt to lock L1 | Attempt to lock L2
t2   |    Gets lock L1    |    Gets lock L2
t3   |      In cs L1      |      In cs L2 
t4   | Attempt to lock L2 | Attempt to lock L1
t5   |     Block on L2    |     Block on L1
<waits forever: deadlock> | <waits forever: deadlock>
```

#### Anything else we need to know?:
You can read more on ABBA deadlock

### PR DESCRIPTION
**Pull request by bvbfan**
/kind fix

#### What this PR does / why we need it:

It ensures atomic lock on multiple mutexes

Fixes #368 #458

### CODE REVIEW

File: src/test/sync_tests.cpp [hunk 1]
```diff
@@ -5,48 +5,49 @@
...
+        TryPotentialDeadlock(rmutex1, rmutex2);
+    }
 
-    #ifdef DEBUG_LOCKORDER
-    g_debug_lockorder_abort = prev;
-    #endif
+    {
+        Mutex mutex1, mutex2;
+        TryPotentialDeadlock(mutex1, mutex2);
+    }
```
Comment (monstrobishi) [inline src/test/sync_tests.cpp:50]: Does this test reproduce the issue this fix is trying to solve

Comment (bvbfan) [reply src/test/sync_tests.cpp:50]: Yes, that's the issue,

**Comment by uzyn** [comment]
Need reviewers

**Comment by uzyn** [comment]
@bvbfan let's rebase this to master and not 1.7

**Comment by bvbfan** [comment]
Let's keep open till 1.7.x is merged to master, then i will rebase it to master.

**Review by monstrobishi — APPROVED** [review]

**Pull request merged**